File: GRMCore/Class/cProjectBAK.vb
Imports System.Runtime.Serialization.Formatters.Binary
Public Class cProjectBAK
    Implements ICloneable
    Public mCVs As List(Of cCVAttribute)
    Public mWatchPoint As cSetWatchPoint
    Public mFCGrid As cFlowControl

    Public Sub New()
        mCVs = New List(Of cCVAttribute)
        mWatchPoint = New cSetWatchPoint
        mFCGrid = New cFlowControl
    End Sub

    Public Sub SetCloneUsingCurrentProject(project As cProject) 'As cProjectBAK
        mCVs = project.CVs
        mWatchPoint = project.WatchPoint
        mFCGrid = project.FCGrid
        Me.Clone()
    End Sub

    Public ReadOnly Property CVs As List(Of cCVAttribute)
        Get
            Return mCVs
        End Get
    End Property

    Public ReadOnly Property WatchPoint As cSetWatchPoint
        Get
            Return mWatchPoint
        End Get
    End Property

    Public ReadOnly Property FCGrid As cFlowControl
        Get
            Return mFCGrid
        End Get
    End Property

    Public ReadOnly Property CV(index As Integer) As cCVAttribute
        Get
            Return mCVs(index)
        End Get
    End Property

    Public Function Clone() As Object Implements ICloneable.Clone
        Dim clo As New cProjectBAK
        For Each cv As cCVAttribute In Me.mCVs
            clo.mCVs.Add(CType(cv.Clone, cCVAttribute))
        Next
        clo.mWatchPoint = CType(Me.WatchPoint.Clone, cSetWatchPoint)
        clo.mFCGrid = CType(Me.FCGrid.Clone, cFlowControl)
        mCVs = clo.mCVs
        mWatchPoint = clo.mWatchPoint
        mFCGrid = clo.mFCGrid
        Return clo
    End Function

End Class

File: GRMCore/Class/cProjectBAK.cs
using System;

namespace GRMCore
{
    public class cProjectBAK : ICloneable
    {
        public cCVAttribute[] CVs;
        public cSetWatchPoint watchPoint;
        public cFlowControl fcGrid;

        public cProjectBAK()
        {
            watchPoint = new cSetWatchPoint();
            fcGrid = new cFlowControl();
        }

        public void SetCloneUsingCurrentProject(cProject project)
        {
            CVs = project.CVs;
            watchPoint = project.watchPoint;
            fcGrid = project.fcGrid;
            Clone();
        }

        public cCVAttribute CV(int index)
        {
            return CVs[index];
        }

        public object Clone()
        {
            cProjectBAK clo = new cProjectBAK();
            clo.CVs = (cCVAttribute[])CVs.Clone();
            clo.watchPoint = (cSetWatchPoint)this.watchPoint.Clone();
            clo.fcGrid = (cFlowControl)this.fcGrid.Clone();
            CVs = clo.CVs;
            watchPoint = clo.watchPoint;
            fcGrid = clo.fcGrid;
            return clo;
        }
    }
}
